pkg/secrets-store/nodeserver.go

```diff
@@ -159,104 +173,116 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
+	providerBinary := ns.getProviderPath(runtime.GOOS, providerName)
+	if _, err := os.Stat(providerBinary); err != nil {
+		errorReason = ProviderBinaryNotFound
+		log.Errorf("failed to find provider %s, err: %v", providerName, err)
+		return nil, err
+	}
 
-		cmd := exec.Command(
-			providerBinary,
-			args...,
-		)
+	parametersStr, err := json.Marshal(parameters)
+	if err != nil {
+		log.Errorf("failed to marshal parameters, err: %v", err)
+		return nil, err
+	}
+	secretStr, err := json.Marshal(secrets)
+	if err != nil {
+		log.Errorf("failed to marshal secrets, err: %v", err)
+		return nil, err
+	}
+	permissionStr, err := json.Marshal(permission)
+	if err != nil {
+		log.Errorf("failed to marshal file permission, err: %v", err)
+		return nil, err
+	}
 
-		stdout := &bytes.Buffer{}
-		stderr := &bytes.Buffer{}
-		cmd.Stderr, cmd.Stdout = stderr, stdout
+	// mount before providers can write content to it
+	err = ns.mounter.Mount("tmpfs", targetPath, "tmpfs", []string{})
+	if err != nil {
+		errorReason = FailedToMount
+		log.Errorf("mount err: %v", err)
+		return nil, err
+	}
 
-		err = cmd.Run()
+	log.Debugf("Calling provider: %s", providerName)
 
-		log.Infof(string(stdout.String()))
-		if err != nil {
-			ns.mounter.Unmount(targetPath)
-			log.Errorf("error invoking provider, err: %v, output: %v", err, stderr.String())
-			return nil, fmt.Errorf("error mounting secret %v", stderr.String())
-		}
+	// check if minimum compatible provider version with current driver version is set
+	// if minimum version is not provided, skip check
+	if _, exists := ns.minProviderVersions[providerName]; !exists {
+		log.Warningf("minimum compatible %s provider version not set", providerName)
 	} else {
-		// mock provider is used only for running sanity tests against the driver
-		err := ns.mounter.Mount("tmpfs", targetPath, "tmpfs", []string{})
+		// check if provider is compatible with driver
+		providerCompatible, err := version.IsProviderCompatible(providerBinary, ns.minProviderVersions[providerName])
 		if err != nil {
-			log.Errorf("mount err: %v", err)
 			return nil, err
 		}
-		log.Infof("skipping calling provider as its mock")
+		if !providerCompatible {
+			errorReason = IncompatibleProviderVersion
+			return nil, fmt.Errorf("Minimum supported %s provider version with current driver is %s", providerName, ns.minProviderVersions[providerName])
+		}
+	}
+
+	args := []string{
+		"--attributes", string(parametersStr),
+		"--secrets", string(secretStr),
+		"--targetPath", string(targetPath),
+		"--permission", string(permissionStr),
+	}
+
+	log.Infof("provider command invoked: %s %s %v", providerBinary,
+		"--attributes [REDACTED] --secrets [REDACTED]", args[4:])
+
+	cmd := exec.Command(
+		providerBinary,
+		args...,
+	)
+
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+	cmd.Stderr, cmd.Stdout = stderr, stdout
+
+	err = cmd.Run()
+
+	log.Infof(string(stdout.String()))
+	if err != nil {
+		errorReason = ProviderError
```

Comment:
Maybe we can parse the error for common errors to match specific errorReason:
`AccessDenied`, `403`, `permission denied`  : `ProviderAccessDenied`
`connection timed out` : `ProviderConnectionTimedOut`